### PR TITLE
Close/Revoke CarePlans and Associated Tasks Using Configured PlanDefinitions

### LINKED ISSUE #2168: Closing/Revoking CarePlans and associated Tasks using configured PlanDefinitions
**Describe the feature request.**
Extend the current event handling system to handle configuring a Questionnaire to close a CarePlan and all the CarePlan's associated Tasks when the Questionnaire is submitted.

**Implementation plan (For Engineers)**
- [ ] Create a `CarePlanConfig` data class that has the following fields
- Conditional FHIRPath expression
- ~CarePlan operation - An enum with the values `close` and `complete`~
- [ ] Update the [QuestionnaireConfig](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/QuestionnaireConfig.kt) to have a new field that holds a list of `CarePlanConfig` objects
- [ ] Create a function that 
- Takes QuestionnaireConfig as a parameter
- Iterates through the list of PlanDefinitions. For each PlanDefinition 
   - Fetch the corresponding CarePlan. Example can be found [here](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt#L95-L98)
   - Evaluate the conditional FHIRPath expression found in the CarePlan config. The resource to use for this is referenced in the `QuestionnaireResponse.subject` field
   - If the FHIRPath expression evaluates to false then terminate execution at this point
   - If the FHIRPath expression evaluates to true then continue to the steps outlined below
       - Update CarePlan status to ~`revoked`~ `completed`
       - Fetch all Tasks linked to the CarePlan. Use the `CarePlan.activity.outcomeReference`
       - Then update all the linked Tasks to `cancelled` that are not 'complete'
- [ ] invoke the function in the [extractCarePlan](https://github.com/opensrp/fhircore/blob/main/android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt#L337-L360) function after the `fhirCarePlanGenerator.generateOrUpdateCarePlan` call

**Acceptance criteria**
The following are examples of workflows that would trigger closing of a CarePlan and associated Tasks
- [ ] An ANC patient will have a pregnancy outcome form. When this form is submitted with a pregnancy outcome of any type the ANC CarePlan and Tasks should be canceled since the patient will move to a different workflow with a different CarePlan.
- [ ] If a patient/family is removed through the family or patient removal forms then all the CarePlans and Tasks for the patient/family should be marked as canceled.
- [ ] All child-related CarePlans should be closed if the patient's Age is updated to more than 5yrs.  Current Child CarePlans are Immunization, Child Routine Visit and Sick Child

### PR DESCRIPTION
**Pull request by qiarie**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2168 

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt [hunk 1]
```diff
@@ -297,4 +300,61 @@ constructor(
...
+        return@forEach
+      }
+
+      if (questionnaireConfig.carePlan.first().operation == CarePlanOperation.CLOSE) {
+        carePlan.status = CarePlan.CarePlanStatus.REVOKED
+        fhirEngine.update(carePlan)
+      }
+
+      carePlan
+        .activity
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt:347]: Is it possible to use Kotlin sequence here?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/CarePlanConfig.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.domain.model
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class CarePlanConfig(
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/CarePlanConfig.kt:22]: Can we document this?

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/QuestionnaireConfig.kt [hunk 3]
```diff
@@ -38,7 +39,8 @@ data class QuestionnaireConfig(
   val groupResource: GroupResourceConfig? = null,
   val taskId: String? = null,
   val saveDraft: Boolean = false,
-  val snackBarMessage: SnackBarMessageConfig? = null
+  val snackBarMessage: SnackBarMessageConfig? = null,
+  val carePlan: List<CarePlanConfig> = emptyList()
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/QuestionnaireConfig.kt:43]: think we should change name
```suggestion
  val carePlanConfigs: List<CarePlanConfig> = emptyList()
```

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt [hunk 4]
```diff
@@ -297,4 +300,61 @@ constructor(
...
+    questionnaireConfig.planDefinitions?.forEach { planDefinition ->
+      val carePlan =
+        fhirEngine
+          .search<CarePlan> {
+            filter(
+              CarePlan.INSTANTIATES_CANONICAL,
+              { value = "${PlanDefinition().fhirType()}/$planDefinition" }
+            )
+          }
+          .firstOrNull()
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt:317]: why do we only get the first CarePlan?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt [hunk 5]
```diff
@@ -297,4 +300,61 @@ constructor(
...
+              subject
+            }
+          fhirPathEngine.evaluateToBoolean(null, null, base, carePlanConfig.fhirPathExpression)
+        }
+
+      if (!conditionIsTrue) {
+        return@forEach
+      }
+
+      if (questionnaireConfig.carePlan.first().operation == CarePlanOperation.CLOSE) {
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt:341]: same question about only first carePlan

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/CarePlanConfig.kt (hunk 2 above)
Comment (pld) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/CarePlanConfig.kt:32]: done

**Comment by pld** [comment]
uhh so I kinda just went for it and got this merged based on the discussion for #2169. Figured we're gonna need some tweaks any ways so we can add those to this code in the process of completing #2169.

I made the assumption that for a Questionnaire with multiple CarePlanConfigs, for all CarePlans attached to any valid PlanDef we will test all CarePlanConfigs to see if any of those say we should update the CarePlan. This is certainly not opitmized on the query side or the loop side, that also feels like something we can do in the process of #2169 or after. As long as #2169 closes with both of them functional we can open a new issue for further optimization cleanup on this code (if needed).

**Review by pld — APPROVED** [review]

**Pull request merged**